subsys/bluetooth/audio/tbs.c

```diff
@@ -0,0 +1,2485 @@
+		if (uri[i] == ':') {
+			(void)memcpy(uri_scheme, uri, i);
+		}
+	}
+
+	if (strlen(uri_scheme) == 0) {
+		/* No URI scheme found */
+		return NULL;
+	}
+
+	for (size_t i = 0; i < ARRAY_SIZE(svc_insts); i++) {
+		for (size_t j = 0; j < ARRAY_SIZE(svc_insts[i].calls); j++) {
+			if (uri_scheme_in_list(uri_scheme,
+					       svc_insts[i].uri_scheme_list)) {
+				return &svc_insts[i];
+			}
+		}
+	}
+
+	return NULL;
+}
+
+static struct tbs_service_inst *lookup_inst_by_work(const struct k_work *work)
+{
+	if (work == NULL) {
+		return NULL;
+	}
+
+	for (size_t i = 0; i < ARRAY_SIZE(svc_insts); i++) {
+		if (&svc_insts[i].reporting_interval_work.work == work) {
+			return &svc_insts[i];
+		}
+	}
+
+	return NULL;
+}
+
+static void tbs_set_terminate_reason(struct tbs_service_inst *inst,
+				     uint8_t call_index, uint8_t reason)
+{
+	inst->terminate_reason.call_index = call_index;
+	inst->terminate_reason.reason = reason;
+	BT_DBG("Index %u: call index 0x%02x, reason %s",
+		inst->index, call_index, bt_tbs_term_reason_str(reason));
+
+	bt_gatt_notify_uuid(NULL, BT_UUID_TBS_TERMINATE_REASON,
+			    inst->service_p->attrs,
+			    (void *)&inst->terminate_reason,
+			    sizeof(inst->terminate_reason));
+
+	if (IS_ENABLED(CONFIG_BT_GTBS)) {
+		bt_gatt_notify_uuid(NULL, BT_UUID_TBS_TERMINATE_REASON,
+				    gtbs_inst.service_p->attrs,
+				    (void *)&inst->terminate_reason,
+				    sizeof(inst->terminate_reason));
+	}
+}
+
+/**
+ * @brief Gets the next free call_index
+ *
+ * For each new call, the call index should be incremented and wrap at 255.
+ * However, the index = 0 is reserved for outgoing calls
+ *
+ * @return uint8_t The next free call index
+ */
+static uint8_t next_free_call_index(void)
+{
+	for (int i = 0; i < CONFIG_BT_TBS_MAX_CALLS; i++) {
+		static uint8_t next_call_index = 1;
+		const struct bt_tbs_call *call = lookup_call(next_call_index);
+
+		if (call == NULL) {
+			return next_call_index++;
+		}
+
+		next_call_index++;
+		if (next_call_index == UINT8_MAX) {
+			/* call_index = 0 reserved for outgoing calls */
+			next_call_index = 1;
+		}
+	}
+
+	BT_DBG("No more free call spots");
+
+	return BT_TBS_FREE_CALL_INDEX;
+}
+
+
+static void net_buf_put_call_state(const void *inst_p)
+{
+	const struct bt_tbs_call *call;
+	const struct bt_tbs_call *calls;
+	size_t call_count;
+
+	net_buf_simple_reset(&read_buf);
+
+	if (inst_p == NULL) {
+		return;
+	}
+
+	if (IS_ENABLED(CONFIG_BT_GTBS) && inst_p == &gtbs_inst) {
+		for (size_t i = 0; i < ARRAY_SIZE(svc_insts); i++) {
+			calls = svc_insts[i].calls;
+			call_count = ARRAY_SIZE(svc_insts[i].calls);
+
+			for (size_t j = 0; j < call_count; j++) {
+				call = &calls[j];
+				if (call->index == BT_TBS_FREE_CALL_INDEX) {
+					continue;
+				}
+				(void)net_buf_simple_add_u8(&read_buf,
+							    call->index);
+				(void)net_buf_simple_add_u8(&read_buf,
+							    call->state);
+				(void)net_buf_simple_add_u8(&read_buf,
+							    call->flags);
```

Comment:
Honestly at this point I don't even know anymore :D 

Some reviewers require that all return values are either explicitly ignored by (void), or otherwise stored in variables, and some say it is okay to leave some out. 

There is a MISRA rule that states that if the function returns error information, that shall be tested (which we also do not do in many places). That is not the case here though. 

Perhaps we should conclude on whether we always shall explicitly handle all return values (e.g. by adding `(void)`, or whether it is okay to not do it), as that is clearly something that reviewers have different opinions on (we also have a bunch of other unwritten rules that are "enforced" by different reviewers in terms of coding style...).

Reply:
The main thing I care about is subsystem-wide consistency. Either this should be done everywhere, or nowhere.

Reply:
We also have the quite useful tool `__must_check` that's currently nowhere nearly sufficiently used in the tree. If it was, I think the importance of these void casts would get reduced (for me they at least have the effect of reduced readability, i.e. they cause me to stop for a sec and wonder "what's going on here").

Reply:
And just to follow up on that: using `__must_check` implies that it's almost always a bug to ignore the return value, and a void cast is therefore mandated to make the ignoring explicit and clearly intentional. In the case of these net_buf APIs the return value is a purely optional convenience that can be used if it makes sense, but in the majority of uses it's not really needed or useful, kind of like an optional input parameter where you can pass NULL (or similar) if you don't care. Therefore forcing the majority of users of the API to do explicit void casts doesn't feel right to me.

I understand where MISRA is coming from with this, but the core problem is that standard C doesn't have a way of declaring optional vs mandatory return values, so the safe bet is to treat everything as mandatory. However, we can do better than that with the help of `__must_check`, and if it was properly used then the forced casts would only happen for APIs where the return value has been annotated as mandatory.

Reply:
I meant per function. If `(void)` is required for a function to ignore its return value in one place, then it should be used everywhere, and vice-versa. Also, I don't think MISRA cares about the meaning of the return value (if it's an error code or something else). As long as there is a return value you must to do something with it according to MISRA (a void cast being one option of "doing something").

> __must_check is a great thing, but should only be used for functions that may return an error value

Why do you say that? I don't think that's accurate. E.g. `net_buf_ref()` doesn't return an error value but uses `__must_check` in order to make sure that there are no reference leaks in the system. Likewise, it'd make sense to apply `__must_check` on any allocator-style functions or functions like `k_fifo_get()` - not just to catch allocation errors, but to make sure that the returned value/memory isn't leaked.

Anyway, this is getting a bit beside the point since the `net_buf_simple_add_*()` APIs don't fall into those categories. As I said, the purpose of the return value there is to be a friendly optional feature that may be used in some rare situations, and forcing the use of `(void)` on the majority of users who don't need this feature doesn't feel quite right.

Reply:
> Also, I don't think MISRA cares about the meaning of the return value (if it's an error code or something else)

The MISRA rule I'm thinking of, is this (from https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html#main-rules)

![image](https://user-images.githubusercontent.com/2520413/145200942-a2e94330-3d2b-4e02-b68e-07e6b98c10ea.png)

> Why do you say that? I don't think that's accurate. E.g. net_buf_ref() doesn't return an error value but uses __must_check in order to make sure that there are no reference leaks in the system.

That's true, it is also very useful for such examples. 

> Anyway, this is getting a bit beside the point since the net_buf_simple_add_*() APIs don't fall into those categories. 

Personally I don't see any reason why we wouldn't do it for `net_buf_simple_add_*` if we do it for e.g. `memset`. I also don't think it harms anything that the code explicitly  ignore the return value, but if you really want it gone, I'll remove it.

Reply:
It might be just my personal way of thinking about explicit casts. But whenever I see them while reading code it makes me stop and think "what's going on here? Is this right?". That's quite wasteful since it's almost always fine to ignore these return values. Anyway, wrt this specific case, there are three acceptable options from my perspective, in order of preference:

1. Don't do the explicit cast.
2. If there are no users of the return value in the tree just make these APIs return void.
3. Do the explicit cast everywhere for these functions.

This PR as it stands now doesn't do any of those options.

Reply:
Regarding that MISRA rule, I'm then confused why explicit casts have been introduced in Zephyr for libc functions that return a value but it's not an error (like memset or memcpy). I wonder if there's some other rule which requires it..

Reply:
> Regarding that MISRA rule, I'm then confused why explicit casts have been introduced in Zephyr for libc functions that return a value but it's not an error (like memset or memcpy). I wonder if there's some other rule which requires it..

Agreed. I figured that someone some time added that requirement, or at least recommendation (we still have some uses of those which aren't explicitly cast), but they do not make any more sense than explicit casts for most of the net_buf functions. 

> This PR as it stands now doesn't follow any of those options. 

I'll remove it for now. Then, as part of the discussion for more general code guidelines/rules for the BT subsystem, we can revisit it.